msg.go

```diff
@@ -962,70 +962,76 @@ func compressedLen(dns *Msg, compress bool) int {
 	return l
 }
 
-func compressionLenSlice(len int, c map[string]int, rs []RR) int {
-	var l int
+func compressionLenSlice(lenp int, c map[string]int, rs []RR) int {
+	initLen := lenp
 	for _, r := range rs {
 		if r == nil {
 			continue
 		}
-		// track this length, and the global length in len, while taking compression into account for both.
+		// TmpLen is to track len of record at 14bits boudaries, 6 bytes is the raw overhead
+		tmpLen := lenp + 6
 		x := r.len()
-		l += x
-		len += x
-
-		k, ok := compressionLenSearch(c, r.Header().Name)
+		// track this length, and the global length in len, while taking compression into account for both.
+		k, ok, sz := compressionLenSearch(c, r.Header().Name)
 		if ok {
-			l += 1 - k
-			len += 1 - k
-		}
-
-		if len < maxCompressionOffset {
-			compressionLenHelper(c, r.Header().Name)
+			x += 1 - k
 		}
+		tmpLen += sz
 
-		k, ok = compressionLenSearchType(c, r)
+		tmpLen += compressionLenHelper(c, r.Header().Name, tmpLen)
+		k, ok, added := compressionLenSearchType(c, r)
 		if ok {
-			l += 1 - k
-			len += 1 - k
-		}
-
-		if len < maxCompressionOffset {
-			compressionLenHelperType(c, r)
+			x += 1 - k
 		}
+		tmpLen += added
+		tmpLen += compressionLenHelperType(c, r, tmpLen)
+		lenp += x
 	}
-	return l
+	return lenp - initLen
 }
 
-// Put the parts of the name in the compression map.
-func compressionLenHelper(c map[string]int, s string) {
+// Put the parts of the name in the compression map, return the size added in payload
+func compressionLenHelper(c map[string]int, s string, currentLen int) int {
+	addedSize := 0
```

Comment:
addedLen ?

Reply:
Well, my comment is probably not clear: compressionLenHelper return the size added to global DNS response (thus the addedSize name).

Considering adding "host1.example.com" with a compression map having already:
"host0.example.com" (and thus "example.com" as well as "com")

it is supposed to return 2+ len("host0")  + 2 (pointer to "example.com" label


Do you have a suggestion to have clearer name?